internal/server/data/access_key.go

```diff
@@ -100,33 +96,88 @@ func CreateAccessKey(db GormTxn, accessKey *models.AccessKey) (body string, err
 	return fmt.Sprintf("%s.%s", accessKey.KeyID, accessKey.Secret), nil
 }
 
-func SaveAccessKey(db GormTxn, key *models.AccessKey) error {
+func UpdateAccessKey(tx WriteTxn, key *models.AccessKey) error {
 	if key.Secret != "" {
 		key.SecretChecksum = secretChecksum(key.Secret)
 	}
+	// TODO: we should perform the same validation as Create
 
-	return save(db, key)
+	return update(tx, (*accessKeyTable)(key))
 }
 
-func ListAccessKeys(db GormTxn, p *models.Pagination, selectors ...SelectorFunc) ([]models.AccessKey, error) {
-	return list[models.AccessKey](db, p, selectors...)
+type ListAccessKeyOptions struct {
+	IncludeExpired bool
+	ByIssuedForID  uid.ID
+	ByName         string
+	Pagination     *Pagination
 }
 
-func GetAccessKey(tx GormTxn, selectors ...SelectorFunc) (*models.AccessKey, error) {
-	db := tx.GormDB()
-	// GetAccessKey by keyID needs to not set an organization_id in the query.
-	// keyID should be globally unique.
-	for _, selector := range selectors {
-		db = selector(db)
-	}
-	result := new(models.AccessKey)
-	if err := db.Model(result).First(result).Error; err != nil {
-		if errors.Is(err, gorm.ErrRecordNotFound) {
-			return nil, internal.ErrNotFound
-		}
+func ListAccessKeys(tx ReadTxn, opts ListAccessKeyOptions) ([]models.AccessKey, error) {
+	table := &accessKeyTable{}
+	query := Query("SELECT")
+	query.B(columnsForSelect("k", table.Columns()))
+	query.B(", u.name")
+	if opts.Pagination != nil {
+		query.B(", count(*) OVER()")
+	}
+	query.B("FROM access_keys AS k INNER JOIN identities AS u")
+	query.B("ON k.issued_for = u.id")
+	query.B("WHERE k.deleted_at is null AND u.deleted_at is null")
+
+	if !opts.IncludeExpired {
+		// TODO: can we remove the need to check for both the zero value and nil?
```

Comment:
we could remove this by setting the extension deadline to the same time as the expiry when the extension deadline isnt specified